### PR TITLE
[FEATURE] Use TIFFTAG_DOCUMENTNAME as layer name if available

### PR DESCRIPTION
**Pull request by oparkins**
## Description

Originally the GDAL data provider ignores any metadata associated with TIFF files. Many documents contain metadata that provides the name and description of the file. This is more important to provide to the user when dealing with sublayers where users are unable to provide a name externally by renaming the file.

This commit allows for sublayers to be defined by this metadata and provided to the user for sublayer selection. If it doesn't exist, it performs the default behavior.

I only modified this for TIFF files with sublayers because in my limited experience, users expect the single layer files to be imported with the same name as found on the operating system. This model breaks down with sublayers since the user cannot provide a sane name for each sublayer.

Related to qgis/QGIS#57077

In the issue mentioned above, I originally stated I didn't think this changed needed to be implemented in the `SUBDATASET` querying code. I changed my approach once I started to dig into this a bit more and found that the `name` was not used as the `uri` for the sublayer location. This enable the change to be made only in one spot and affect the display of the layer elsewhere.

---

Here is what the import originally looks like for a multi-page tiff file:

before_sublayer_dialog

And the default layers list once imported:

before_layers_list

And now:

after_sublayer_dialog

And the default layers list once imported:

after_layers_list

---

Todo:
- [x] Add unit test with sublayers
- [x] Ensure scripts are ran before commits as stated in PR help text
- [x] Check coding standards document as mentioned in the PR help text

For the tests, I just added another example to the repo that mirrored the `gtiff_tags` test. The test that I modified still passes.

### CODE REVIEW

**Comment by oparkins** [comment]
This should be good for someone to look at now. Really appreciate the structure of QGIS and the code quality since it was easy to implement this feature. 

Let me know if there is anything I should fix/change/etc. Looking forward to having this available for our multi-page TIFF files we generate.

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 1]
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
         }
         else
         {
+
+          // Check if the layer has TIFFTAG_DOCUMENTNAME associated with it. If so, use that name.
+          GDALDatasetH datasetHandle = GDALOpen( name, GA_ReadOnly );
```
Comment (rouault) [inline src/core/providers/gdal/qgsgdalprovider.cpp:1840]: below code addition should only be execute under ``if (datasetHandle)``

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 2]
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
...
+          QString tagTIFFDocumentName = GDALGetMetadataItem( datasetHandle, "TIFFTAG_DOCUMENTNAME", nullptr );
+          if ( tagTIFFDocumentName != nullptr )
```
Comment (rouault) [inline src/core/providers/gdal/qgsgdalprovider.cpp:1842]: tagTIFFDocumentName is a QString, not a pointer, so can' be nullptr:

```suggestion
          if ( !tagTIFFDocumentName.isEmpty() )
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 3]
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
...
+          {
+            layerName = tagTIFFDocumentName;
+          }
+
+          QString tagTIFFImageDescription = GDALGetMetadataItem( datasetHandle, "TIFFTAG_IMAGEDESCRIPTION", nullptr );
+          if ( tagTIFFImageDescription != nullptr )
```
Comment (rouault) [inline src/core/providers/gdal/qgsgdalprovider.cpp:1848]:
```suggestion
          if ( !tagTIFFImageDescription.isEmpty() )
```

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 1 above)
Comment (oparkins) [reply src/core/providers/gdal/qgsgdalprovider.cpp:1840]: Good point! Thanks!

**Comment by oparkins** [comment]
Thanks for the review @rouault. I've made those changes.

**Comment by oparkins** [comment]
Just pushed a fix for those tests. Proof of them working on my machine now:

```bash
ctest -R test_core_gdalprovider
Test project /home/oparkins/projects/QGIS/build-master
    Start  1: logGitStatus
1/3 Test  #1: logGitStatus .....................   Passed    0.05 sec
    Start 70: test_core_gdalprovider
2/3 Test #70: test_core_gdalprovider ...........   Passed    1.01 sec
    Start  2: checkGitStatus
3/3 Test  #2: checkGitStatus ...................   Passed    0.04 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   1.14 sec
```

Not sure how I messed that up, but I did. This should have fixed that.

**Comment by oparkins** [comment]
I just synced my fork to make future merges easier. I'm also hoping that will enable the workflows to pass. The last round of failures seem to be caused by another test case that is not relevant to my changes.

**Comment by rouault** [comment]
so we have those failures:
```
 	136 - test_core_networkaccessmanager (Timeout)
	356 - test_app_maptooladdpart (Failed)
```
The first one is likely a random flake (I'm restarting the build to test that), and the second is definitely not due to this pull request, as occuring in master builds too (https://github.com/qgis/QGIS/actions/runs/8913989856/job/24481614072)

**Comment by oparkins** [comment]
Awesome. Thanks for checking into that. Glad I wasn't going too crazy.

Besides those two failed tests, this PR should contain all that is needed to close the associated issue https://github.com/qgis/QGIS/issues/57077. If there is anything that you see needs correction related to this PR, more than happy to keep working on it. Just let me know how I can be helpful.

**Review by rouault — APPROVED** [review]

**Pull request merged**